### PR TITLE
build:  add rsync command + --upload option for linux boards

### PR DESCRIPTION
**Pull request by guludo**
Hi all,

This PR tries to address #6283. It does so by creating a more general `rsync` command and making `linux` boards add it to the command chain when the user passes the `--upload` build option.

One not so good side-effect of this approach is that a command like `waf copter --upload` is executed actually as two commands (`copter` and `rsync`) and the build summary is output twice to the user.

@lucasdemarchi, does that somehow addresses your issue?

Best,
Gustavo Sousa

### CODE REVIEW

File: wscript [hunk 1]
```diff
@@ -10,7 +10,7 @@ sys.path.insert(0, 'Tools/ardupilotwaf/')
 import ardupilotwaf
 import boards
 
-from waflib import Build, ConfigSet, Configure, Context, Utils
+from waflib import Build, ConfigSet, Configure, Context, Options, Utils
```
Comment (guludo) [inline wscript]: Oops... This is a leftover from a previous version...

File: wscript [hunk 2]
```diff
@@ -397,3 +397,18 @@ for program_group in ('all', 'bin', 'tools', 'examples', 'tests', 'benchmarks'):
         program_group_list=program_group,
         doc='builds all programs of %s group' % program_group,
     )
+
+class LocalInstallContext(Build.InstallContext):
+    """runs install using BLD/install as destdir, where BLD is the build variant directory"""
+    cmd = 'linstall'
```
Comment (lucasdemarchi) [inline wscript]: could this be localinstall?

Comment (guludo) [inline wscript]: Yes. I just though it would be too much typing. Maybe local_install then?

**Comment by lucasdemarchi** [comment]
LGTM, thanks!

Small comment above and now I'm thinking if it wouldn't make more sense to have `upload` as a target like you did in your snippet on the issue. What do you think?

File: wscript (hunk 2 above)
Comment (lucasdemarchi) [inline wscript]: As far as I understood the user wouldn't need to run this since it's brought automatically by the upload logic, right?

**Comment by guludo** [comment]
> Small comment above and now I'm thinking if it wouldn't make more sense to have upload as a target like you did in your snippet on the issue. What do you think?

I though of using `rsync` because there could be different ways of uploading and `upload` would be too general then.

**Comment by lucasdemarchi** [comment]
> I though of using rsync because there could be different ways of uploading and upload would be too general then.

Suppose we have 2 upload mechanisms: using rsync for "normal linux boards" and adb for bebop and the like.  Wouldn't it be better to always have an "upload" command to the user and based on the board we figure out if we should use rsync, adb or something else?

File: wscript (hunk 2 above)
Comment (guludo) [inline wscript]: Right. I'll change to it.

**Comment by guludo** [comment]
> Suppose we have 2 upload mechanisms: using rsync for "normal linux boards" and adb for bebop and the like. Wouldn't it be better to always have an "upload" command to the user and based on the board we figure out if we should use rsync, adb or something else?

Hm... Yes, that makes sense. In fact, that's how `--upload` is working now, but that is more like a convenience to the build commands. So, I think we could have an `upload` command that calls the default upload mechanism for the board. The command `rsync` would still exist and be available to the user in case she still wants to explicitly use that.

**Comment by lucasdemarchi** [comment]
> So, I think we could have an upload command that calls the default upload mechanism for the board. The command rsync would still exist and be available to the user in case she still wants to explicitly use that.

Yep, sounds good.

**Comment by lucasdemarchi** [comment]
@guludo ping

**Comment by lucasdemarchi** [comment]
I completely missed the fact that this honors the `--upload` switch and thought there would only be a `rsync` target. Thus I'm fine with it as is. I continued it on

**Comment by guludo** [comment]
@lucasdemarchi Thanks and sorry. Life's been crazy lately :-P

**Pull request closed without merging**